Extensions: Manage - Warning for not using Update

pe7er opened the pull request:
Pull Request to distinguish 3rd party extensions that do NOT support the Joomla Extension Update

### Summary of Changes
The JED lists for all extensions if they support the Joomla Extension Update.
However, you cannot see if the extensions you've installed in your Joomla site support the Joomla Extension Update. Maybe the JED lists some extensions that they support the Update, but you are using older versions in your site that do not support it.

To make it easier to see which extensions support the Joomla Extension Update, I've added a warning for 3rd party extensions that do not support it.

### Testing Instructions
Install a 3rd party extension that uses the Joomla Update, for example
https://db8.eu/download/component/db8-optimize-site

Install this patch.

Go to **Extensions: Manage**. You should see the installed 3rd party extension. 

![screen-1](https://cloud.githubusercontent.com/assets/1217850/26754076/ceeee2e2-4873-11e7-969f-e34586ee4560.png)


Go to **Extensions: Update Sites**. **Remove the reference** for the installed 3rd party extension.


![screen-2](https://cloud.githubusercontent.com/assets/1217850/26754078/cef07bac-4873-11e7-9b8a-7e159589a491.png)



![screen-3](https://cloud.githubusercontent.com/assets/1217850/26754077/ceef85a8-4873-11e7-9b8d-32dd3d59799a.png)

Go to **Extensions: Manage**. You should see the installed 3rd party extension plus a message that it does not support Joomla update

![screen-4](https://cloud.githubusercontent.com/assets/1217850/26754079/cef0b8ec-4873-11e7-9013-7dd6e40045b6.png)


The extension did support the Joomla Update, but the reference is missing in the Update Sites and not checked anymore. To enable it again: 
Go to **Extensions: Update Sites**. Choose Rebuild.

Go to **Extensions: Manage**. You should see the installed 3rd party extension without the warning message.

pe7er committed:
Extensions: Manage - warning for not using Update

brianteeman commented on administrator/components/com_installer/models/manage.php:
```diff
@@ -344,7 +346,7 @@ protected function getListQuery()
 
 		// Note: The search for name, ordering and pagination are processed by the parent InstallerModel class (in extension.php).
-
+//echo $query;
```
can you remove this comment

pe7er replied:
Thanks @brianteeman !

brianteeman commented on administrator/language/en-GB/en-GB.com_installer.ini:
```diff
@@ -248,6 +248,8 @@ COM_INSTALLER_UNINSTALL_ERROR="Error uninstalling %s."
 ; This string is deprecated and will be removed with 4.0.
 COM_INSTALLER_UNINSTALL_LANGUAGE="A language should always have been installed as a package. <br />To uninstall a language, filter type by package and uninstall the package."
 COM_INSTALLER_UNINSTALL_SUCCESS="Uninstalling the %s was successful."
+COM_INSTALLER_UPDATE_NOT_SUPPORTED="Update server unavailable"
```
Is it update server not working eg down and broken or is it there is no update server for this extension?

pe7er replied:
Technical explanation:
There is no update server in the manifest of this extension, or there is an update server in the manifest of this extension but someone has removed it in Update Servers.

What would be the best not-too-technical description for normal users?

richard67 replied:
Maybe "Update site not available" is better, because the feature is called "Update sites" in Joomla?

pe7er committed:
Extensions: Manage - warning for not using Update

pe7er committed:
Merge branch 'staging' into supportsupdate

Quy commented on administrator/components/com_installer/models/manage.php:
```diff
@@ -292,9 +292,11 @@ public function remove($eid = array())
 		$query = $this->getDbo()->getQuery(true)
-			->select('*')
+			->select('e.*')
+            ->select('u.update_site_id')
```
Fix indentation. Use tabs.

Quy commented on administrator/components/com_installer/models/manage.php:
```diff
@@ -292,9 +292,11 @@ public function remove($eid = array())
 			->select('2*protected+(1-protected)*enabled AS status')
-			->from('#__extensions')
+			->from('#__extensions AS e')
+            ->leftJoin('#__update_sites_extensions AS u ON u.extension_id = e.extension_id')
```
Fix indentation. Use tabs.

Quy commented on administrator/components/com_installer/views/manage/tmpl/default.php:
```diff
@@ -103,6 +103,12 @@
 									<?php echo $item->name; ?>
 								</span>
 							</label>
+                            <?php if(($item->extension_id > 999) && ($item->update_site_id === null)): ?>
```
Fix indentation. Use tabs. 
Add a space after `if`. 
Remove unnecessary parentheses and add space before `:`.
 `if ($item->extension_id > 999 && $item->update_site_id === null) :`

Quy commented on administrator/language/en-GB/en-GB.com_installer.ini:
```diff
@@ -248,6 +248,8 @@ COM_INSTALLER_UNINSTALL_ERROR="Error uninstalling %s."
 COM_INSTALLER_UNINSTALL_LANGUAGE="A language should always have been installed as a package. <br />To uninstall a language, filter type by package and uninstall the package."
 COM_INSTALLER_UNINSTALL_SUCCESS="Uninstalling the %s was successful."
+COM_INSTALLER_UPDATE_NOT_SUPPORTED="Update server unavailable"
+COM_INSTALLER_UPDATE_NOT_SUPPORTED_DESC="Check Extensions &gt; Update Sites &gt; and use Rebuild button."
```
Add `Manage` after `Extensions >`.
Remove `&gt;` before `and use`.

pe7er committed:
code style + better label

pe7er committed:
Merge branch 'supportsupdate' of https://github.com/pe7er/joomla-cms into supportsupdate

pe7er commented:
Oops, I used tabs but my reinstalled IDE converted them to non visible spaces :-(
Thank you @Quy + @richard67 for your comments!
I've made all the changes that you've recommended.

Quy commented on administrator/components/com_installer/views/manage/tmpl/default.php:
```diff
@@ -103,7 +103,7 @@
 								</span>
 							</label>
-                            <?php if(($item->extension_id > 999) && ($item->update_site_id === null)): ?>
+                            <?php if ($item->extension_id > 999 && $item->update_site_id === null) : ?>
```
Fix indentation/tabs from lines 106 to 111.

Quy commented on administrator/language/en-GB/en-GB.com_installer.ini:
```diff
@@ -248,8 +248,8 @@ COM_INSTALLER_UNINSTALL_ERROR="Error uninstalling %s."
 COM_INSTALLER_UNINSTALL_SUCCESS="Uninstalling the %s was successful."
-COM_INSTALLER_UPDATE_NOT_SUPPORTED="Update server unavailable"
-COM_INSTALLER_UPDATE_NOT_SUPPORTED_DESC="Check Extensions &gt; Update Sites &gt; and use Rebuild button."
+COM_INSTALLER_UPDATE_NOT_SUPPORTED="Update Sites not unavailable"
```
You probably mean `not available`.
Change `SUPPORTED` to `AVAILABLE` in keys.
Move these 2 keys after `COM_INSTALLER_UPDATE_FILTER_SEARCH_LABEL`.

zero-24 committed:
spaces to tabs

zero-24 commented on administrator/components/com_installer/views/manage/tmpl/default.php:
```diff
@@ -103,6 +103,12 @@
 									<?php echo $item->name; ?>
 								</span>
 							</label>
+							<?php if ($item->extension_id > 999 && $item->update_site_id === null) : ?>
```
i don't get the point of `$item->extension_id > 999` ?

wojsmol replied:
@zero-24 IMHO @pe7er wont to exclude Joomla! core extensions

zero-24 replied:
Ah. This makes sense. Thanks

pe7er replied:
Yes, I am indeed excluding the Joomla core extensions with $item->extension_id > 999

richard67 replied:
Hmm, I have found some other PR which checks against 9999 and not 999: #16494 , see this line: [https://github.com/joomla/joomla-cms/pull/16494/files#diff-35b54662480daa18b046886a634f00d6R1016](https://github.com/joomla/joomla-cms/pull/16494/files#diff-35b54662480daa18b046886a634f00d6R1016)
@wilsonge Can you check and let us know which is the right one, 999 or 9999?

AlexRed commented:
Good idea, but if an extension is a Package the update sites is in the package for all the extension part (plugin, component, module ecc...) , but are reported with no update sites like the language package:
![pack-update-sites](https://cloud.githubusercontent.com/assets/1362388/26759723/7745a720-48fe-11e7-9933-8bef04236f57.png)
or JCE or Akeeba Backup

richard67 commented on administrator/language/en-GB/en-GB.com_installer.ini:
```diff
@@ -248,6 +248,8 @@ COM_INSTALLER_UNINSTALL_ERROR="Error uninstalling %s."
 ; This string is deprecated and will be removed with 4.0.
 COM_INSTALLER_UNINSTALL_LANGUAGE="A language should always have been installed as a package. <br />To uninstall a language, filter type by package and uninstall the package."
 COM_INSTALLER_UNINSTALL_SUCCESS="Uninstalling the %s was successful."
+COM_INSTALLER_UPDATE_NOT_SUPPORTED="Update Sites not unavailable"
```
Hmm, I meant "Update Site not available" or "Update Site unavailable". But not "Update Site not unavailable". "**Not un**available" means "Available", so it is not what you want, I guess ;-) Please note also the singular in my proposal: For this particular component, the one and only update site is unavailable. Not all update sites are unavailable. Or do some extensions have more than 1 update site? In this case, the plural would be correct.

Quy replied:
Also put the keys in alpha order.

franz-wohlkoenig commented:
@pe7er is this PR ready for test?

richard67 commented:
@wilsonge What is correct now to check for core extensions? Item ID 999 like done in this PR here? Or 9999 like done in PR #16494 ?
I ask you because the other PR seems to be based on your input.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 commented:
@pe7er Maybe you should change the 999 to 9999 for check of the extension id, see [https://github.com/joomla/joomla-cms/pull/16494#issuecomment-306919727](https://github.com/joomla/joomla-cms/pull/16494#issuecomment-306919727 "https://github.com/joomla/joomla-cms/pull/16494#issuecomment-306919727").<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 commented:
The sidenote in the [comment I mentioned above](https://github.com/joomla/joomla-cms/pull/16494#issuecomment-306919727 "https://github.com/joomla/joomla-cms/pull/16494#issuecomment-306919727") makes me think it is not really save to use the ID number to check if an extension is core or not. But what else way do we have beside a hard-coded list which would have to be maintained every time an extension is added (and so would be a shit)?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 commented:
Regarding the question on how to detect if an extension is core or not which I mentioned in my comment above (999 or 9999), this PR here should not wait for a solution.

Let it continue to use the 999.

Any later PR implementing a safe check will also have to look for such things and replace them by using the new way which it implements.

But this PR needs 2 changes to be good:

1. Change text COM_INSTALLER_UPDATE_NOT_SUPPORTED from "Update Sites not **un**available" to "Update Sites not available".
2. Change the database query so it excludes extensions where the package id is set and which are not of type package from the check, so for these no warning is shown, or alternatively use the update site of the package for checking those extensions so the status of the update site of their package is shown.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 reviewed (changes requested):
Regarding the question on how to detect if an extension is core or not which I mentioned in my comment above (999 or 9999), this PR here should not wait for a solution.

Let it continue to use the 999.

Any later PR implementing a safe check will also have to look for such things and replace them by using the new way which it implements.

But this PR needs 2 changes to be good:

    Change text COM_INSTALLER_UPDATE_NOT_SUPPORTED from "Update Sites not unavailable" to "Update Sites not available".
    Change the database query so it excludes extensions where the package id is set and which are not of type package from the check, so for these no warning is shown, or alternatively use the update site of the package for checking those extensions so the status of the update site of their package is shown.

AndyGaskell commented:
This sounds a really nice feature, couple of very minor comments on the "Update Sites not unavailable" text....

1. I would think perhaps the sites should not be capitalised, so "Update sites not unavailable".
2. Are there usually multiple update sites? I've only ever set up a single update site, I thought that was the normal way to do it.  In which case it should perhaps be "site" rather than "sites", so   "Update site not unavailable".

These are very minor comments, and no problem at all if folk disagree.

Thanks :)

richard67 commented:
Does nobody here notice that "not unavailable" is a mistake and it should either be "unavailable" or "not available"??? Do you sleep while reading?

AndyGaskell commented:
I think @richard67 that you're probably right, may have been sleep reading there.  

I'd say, on reflection, that double negatives tend to be confusing, so you're right, "unavailable" or "not available" would be better wording.  

Also, wording aside "not unavailable" is, I guess not the intended description.

richard67 commented:
@AndyGaskell Well, at the beginning the text was "unavailable", but to me "not available" sounded better so I suggested that, and @pe7er wanted to follow my suggestion but maybe missunderstood me or understood me right but made a mistake, and so he changed it to "not unavailable", but this negates the original meaning and so displays wrong information.

Quy commented:
@richard67 You can submit a PR to his branch to be merged.

richard67 commented:
I know, if he wants I can do that later.

richard67 commented:
I will have to do that anyway if my PR with the new extensions helper class will be merged before this one.

richard67 committed:
Correct text COM_INSTALLER_UPDATE_NOT_SUPPORTED

Changed "not unavailable" to "not available" so it makes the desired sense.

richard67 commented:
@pe7er I created PR no. 1 in your repository for your branch "supportsupdate". Please check and accept.

pe7er committed:
Merge pull request #19 from richard67/patch-1

Correct text COM_INSTALLER_UPDATE_NOT_SUPPORTED

richard67 commented:
@pe7er What remains to be done in my point of view is what @AlexRed wrote in his comment above:
If the extensions belong to a package, which means the "Package ID" column is not empty, the status of the package's update site has to be shown and not the one of the extension. This can be done by modifying the SQL select statement in file administrator/components/com_installer/models/manage.php by joining the update_sites_extensions table a second time, this time to the package_id column, and then checking the update_site_id of the package if it is set, otherwise the one of the extension. I am working on a PR to your repo for that.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 commented:
@pe7er Another thing is the check to exclude core extensions. I think it either is not necessary at all because the only core extensions which have update sites are those which can be excluded like it is done here: [https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_installer/models/updatesites.php#L382](https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_installer/models/updatesites.php#L382 "https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_installer/models/updatesites.php#L382").

So you can say in your query you exclude those (file/element): 'file'/'joomla', 'package'/'pkg_en-GB', 'component'/'com_joomlaupdate'.

I will try to do that in my PR to your repo which I mentioned in my previous comment.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 committed:
Add parent package's update site ID to list query

Join the update site id of the parent package, too.

richard67 committed:
No warning if parent package has update site

Show update not supported warning only if there is neither an update
site for the extension itself nor for a parent package.

richard67 commented:
@pe7er Check PR [https://github.com/pe7er/joomla-cms/pull/20](https://github.com/pe7er/joomla-cms/pull/20 "https://github.com/pe7er/joomla-cms/pull/20") against your repo. It solves the problem mentioned by @AlexRed .

The other thing with core extensions check I will see what I can do.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 commented:
@pe7er Regarding the core extensions I think it is really necessary to have my PR #16724 merged an then use it instead of checking for extention_id > 999. But for now I think it would be sufficient if you just change the 999 to 9999 so it fits to the auto increment of the extensions table (which starts at 10000 so everything below is a core extension for sure, and only core extensions which have been discovered by the discovery installer function will have larger IDs). I will add that as a commit to my previously mentioned PR to your repo.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 committed:
Check for 9999 and not 999

Check for 9999 and not 999 because auto increment of extension ID starts
with 10000

richard67 committed:
Use safe method for check if not core extension

Use the new JExtensionHelper class instead of the unsafe check for extension_id > 999 for checking if not core extension.

richard67 commented:
@pe7er I added now a commit to my PR for your repository which replaces the unsafe check for extension_id > 999 by using the new JExtensionHelper class.

I haven't tested it yet here, but I will soon and let you know about the result. Of course if you test faster, you are welcome.

richard67 committed:
Correct 2nd join for parent package update site

Correct silly mistake: In the 2nd join I should join to the right table
column (package_id) and not again to the extension_id like the 1st join
already does.

richard67 commented:
@pe7er Meanwhile I have tested my changes in my PR for your repo, and I corrected a silly mistake. Now it is ok. Core extensions are excluded now also if they have extension ID greater than 999 (or more), and for extensions which have a parent package which has an update site (e.g. German language pack), the warning is not shown.

So if you accept and merge my PR into your repo, I can test this one here with success.

But you might have to fix code style of my PR: The line with the condition in the view for check if the warning has to be displayed is a bit long after my changes, so you may have to do some line break.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

richard67 committed:
Change order of checks in if condition

Change order of checks in if condition so the easy checks are done first and the (more cost expensive) check if core extension is done as last (assuming PHP prcesses the conditions from left to right).

brianteeman commented:
@pe7er have you merged the changes from @richard67

pe7er committed:
Merge pull request #20 from richard67/patch-1

1. Check also the parent package's update site, and 2. use safe method to check if not core extension

pe7er commented:
@brianteeman I just merged @richard67 his code. Sorry for my delay!

brianteeman commented:
@pe7er thanks - lets hope we can get it tested and merged now

Quy commented on administrator/components/com_installer/views/manage/tmpl/default.php:
```diff
@@ -103,6 +103,12 @@
 								</span>
 							</label>
+							<?php if ($item->update_site_id === null && $item->update_site_id_pkg === null && JExtensionHelper::checkIfCoreExtension($item->type, $item->element, $item->client_id, $item->folder) === false) : ?>
+								<span class="badge badge-important bold hasTooltip"
```
Remove `bold` class since `font-weight: bold;` is already specified in the `badge` class.

```
.badge {
	display: inline-block;
	padding: 2px 4px;
	font-size: 10.998px;
	font-weight: bold;
	line-height: 14px;
	color: #fff;
	vertical-align: baseline;
	white-space: nowrap;
	text-shadow: 0 -1px 0 rgba(0,0,0,0.25);
	background-color: #999;
}
```

Quy commented on administrator/language/en-GB/en-GB.com_installer.ini:
```diff
@@ -248,6 +248,8 @@ COM_INSTALLER_UNINSTALL_ERROR="Error uninstalling %s."
 ; This string is deprecated and will be removed with 4.0.
 COM_INSTALLER_UNINSTALL_LANGUAGE="A language should always have been installed as a package. <br />To uninstall a language, filter type by package and uninstall the package."
 COM_INSTALLER_UNINSTALL_SUCCESS="Uninstalling the %s was successful."
+COM_INSTALLER_UPDATE_NOT_SUPPORTED="Update Sites not available"
```
Please sort in alpha order. Move lines 251-252 after line 254.

richard67 commented:
I've just tested and noticed following:

1. If an update site is existing but disabled, it is not counted as missing, i.e. the badge is not shown in this case. It is only shown if the update site is really missing in database. @pe7er Is this by purpose? The testing instructions read as if it was by purpose. If so, then this observation is not an issue.

2. For components like e.g. OSMap which come with a component and a plugin and these do not belong to any parent package, this PR does not work. But I think that's by purpose, because I do not see any chance to find out in database that the OSMap content plugin is shipped with the OSMap component. The only chance would be to parse the component's manifest XML. Or do I miss something? If I don't miss something, then this 2nd observation is not an issue.

If both these observations turn out to be ok, i.e. not an issue with this PR, then I can mark my test result positive.

But because of the 2nd observation I think the badges intruduced with this PR could confuse the user, as it confused me at a first view.

@pe7er @brianteeman @Quy @AlexRed @AndyGaskell what do you think?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

wojsmol commented:
@richard67 Regarding your observation number 2. This can be determined by checking value of `package_id` in `#__extensions` table.

richard67 commented:
@wojsmol No, as I wrote, the OSMap stuff is organised differently, it does not use the package_id for its content plugin. What you write, checking the package_id column, is already implemented in this PR here (and I was the one who suggested that change, see history above).

richard67 commented:
@wojsmol Anyway, if there are not more components having the same problem as OSMap and its content plugin, then it is an OSMap issue, and this PR is ok regarding that. But maybe there are other components which do it the same way, I don't know.

richard67 commented:
I have tested this item :white_check_mark: successfully on 72fe81b328dbedc4f3b8f35cd6afbaab5fb2bd81<br /><br />This PR works as described.

What it does not do is to show the badge if an update site is present in database but is disabled. It only shows the badge if the update site is missing in database. But I think this is as intended by this PR, so I do not count this as an error.

What this PR also does not do is not show the badge e.g. for plugins or modules which are installed with a component and not with a package, so the package_id column is not set in the extensions table in database, but the plugin or module is updated together with the component by the online update, like it is e.g. with the OSMap Component and its content plugin. But this is also not within the scope of this PR as I understand it, and so I don't count this as an error.

On the other hand, always showing the badge for those modules or plugins could confuse people and so maybe cause support requests in forums and issues in the tracker. The only chance to proper handle such stuff would be to parse the component's manifest XML, as far as I see. Or we can make component creators make their stuff be delivered in a package or set the package_id column at least.

Maybe the PLT or some regular maintainers should have a look on it and decide if that is ok or not.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16474">issues.joomla.org/tracker/joomla-cms/16474</a>.</sub>

mbabker commented:
> Or we can make component creators make their stuff be delivered in a package or set the package_id column at least.

Extension developers can't be forced to repackage their extensions because of us.  We can't account for every possible distribution model nor should we, so this scenario can be thrown into a list of known issues with a "won't fix" label.

mbabker commented:
Rethinking this it actually feels like it could be problematic or send mixed signals for extensions which are one off custom solutions for a site or client.  For the `joomla.org` backend we'd have several sites where this error badge would show up in error but it's not really a problem because these are our extensions and we aren't publishing update servers for them.  There's a part of me that almost wants to suggest closing this without accepting it between that and the list of potential issues with how different extensions are packaged.

If there's an alternative that doesn't make it come across as an extension is broken or otherwise bad because it doesn't have an update site, that should be considered.  But looking fresh on this I honestly don't feel like this is a good idea.
